File: src/page/anim/EditHead.java
package page.anim;

import common.CommonStatic.EditLink;
import common.util.anim.AnimCE;
import io.BCUWriter;
import page.JBTN;
import page.Page;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class EditHead extends Page implements EditLink {

	private static final long serialVersionUID = 1L;

	private final JBTN undo = new JBTN(0, "undo");
	private final JBTN save = new JBTN(0, "save");
	private final JBTN view = new JBTN(0, "vdiy");
	private final JBTN icut = new JBTN(0, "caic");
	private final JBTN mmdl = new JBTN(0, "camm");
	private final JBTN manm = new JBTN(0, "cama");

	protected AnimCE focus;

	private DIYViewPage p0;
	private ImgCutEditPage p1;
	private MaModelEditPage p2;
	private MaAnimEditPage p3;
	private int val;
	private Page cur;
	private boolean changing = false;

	protected AnimCE anim;

	public EditHead(Page p, int v) {
		super(p.getFront());
		val = v;
		cur = p;
		focus = null;
		if (v == 0)
			p0 = (DIYViewPage) p;
		else if (v == 1)
			p1 = (ImgCutEditPage) p;
		else if (v == 2)
			p2 = (MaModelEditPage) p;
		else if (v == 3)
			p3 = (MaAnimEditPage) p;
		ini();
	}

	@Override
	public void review() {
		undo.setEnabled(anim != null && anim.history.size() > 1);
		cur.callBack("review");
		if (anim != null)
			undo.setToolTipText(anim.getUndo());
	}

	public void setAnim(AnimCE da) {
		if (changing)
			return;
		anim = da;
		if (anim != null)
			anim.link = this;
		review();
	}

	@Override
	protected void resized(int x, int y) {
		set(undo, x, y, 250, 0, 200, 50);
		set(save, x, y, 500, 0, 200, 50);
		set(view, x, y, 800, 0, 200, 50);
		set(icut, x, y, 1050, 0, 200, 50);
		set(mmdl, x, y, 1300, 0, 200, 50);
		set(manm, x, y, 1550, 0, 200, 50);
	}

	private void addListeners() {
		EditHead thi = this;

		view.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent arg0) {
				if (val == 0)
					return;
				changing = true;
				cur.remove(thi);
				if (p0 == null)
					p0 = new DIYViewPage(getFront(), thi);
				changePanel(cur = p0);
				cur.add(thi);
				((AbEditPage) cur).setSelection(anim);
				val = 0;
				changing = false;
			}

		});

		icut.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent arg0) {
				if (val == 1)
					return;
				changing = true;
				cur.remove(thi);
				if (p1 == null)
					p1 = new ImgCutEditPage(getFront(), thi);
				changePanel(cur = p1);
				cur.add(thi);
				((AbEditPage) cur).setSelection(anim);
				val = 1;
				changing = false;
			}

		});

		mmdl.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent arg0) {
				if (val == 2)
					return;
				changing = true;
				cur.remove(thi);
				if (p2 == null)
					p2 = new MaModelEditPage(getFront(), thi);
				changePanel(cur = p2);
				cur.add(thi);
				((AbEditPage) cur).setSelection(anim);
				val = 2;
				changing = false;
			}

		});

		manm.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent arg0) {
				if (val == 3)
					return;
				changing = true;
				cur.remove(thi);
				if (p3 == null)
					p3 = new MaAnimEditPage(getFront(), thi);
				changePanel(cur = p3);
				cur.add(thi);
				((AbEditPage) cur).setSelection(anim);
				val = 3;
				changing = false;
			}

		});

		save.setLnr((e) -> BCUWriter.writeData());

		undo.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent arg0) {
				anim.restore();
				review();
				((AbEditPage) cur).setSelection(anim);
			}

		});

	}

	private void ini() {
		add(view);
		add(icut);
		add(mmdl);
		add(manm);
		add(save);
		add(undo);
		addListeners();
	}

}

File: src/page/anim/EditHead.kt
package page.anim

import common.CommonStatic.EditLink
import common.util.anim.AnimCE
import io.BCUWriter
import page.JBTN
import page.Page
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.util.function.Consumer

class EditHead(private var cur: Page, private var `val`: Int) : Page(cur.front), EditLink {
    private val undo: JBTN = JBTN(0, "undo")
    private val save: JBTN = JBTN(0, "save")
    private val view: JBTN = JBTN(0, "vdiy")
    private val icut: JBTN = JBTN(0, "caic")
    private val mmdl: JBTN = JBTN(0, "camm")
    private val manm: JBTN = JBTN(0, "cama")
    var focus: AnimCE? = null
    private var p0: DIYViewPage? = null
    private var p1: ImgCutEditPage? = null
    private var p2: MaModelEditPage? = null
    private var p3: MaAnimEditPage? = null
    private var changing = false
    var anim: AnimCE? = null
    override fun review() {
        undo.isEnabled = anim != null && anim.history.size > 1
        cur.callBack("review")
        if (anim != null) undo.setToolTipText(anim.getUndo())
    }

    fun setAnim(da: AnimCE?) {
        if (changing) return
        anim = da
        if (anim != null) anim.link = this
        review()
    }

    override fun resized(x: Int, y: Int) {
        Page.Companion.set(undo, x, y, 250, 0, 200, 50)
        Page.Companion.set(save, x, y, 500, 0, 200, 50)
        Page.Companion.set(view, x, y, 800, 0, 200, 50)
        Page.Companion.set(icut, x, y, 1050, 0, 200, 50)
        Page.Companion.set(mmdl, x, y, 1300, 0, 200, 50)
        Page.Companion.set(manm, x, y, 1550, 0, 200, 50)
    }

    private fun addListeners() {
        val thi = this
        view.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                if (`val` == 0) return
                changing = true
                cur.remove(thi)
                if (p0 == null) p0 = DIYViewPage(front, thi)
                changePanel(p0.also { cur = it })
                cur.add(thi)
                (cur as AbEditPage).setSelection(anim)
                `val` = 0
                changing = false
            }
        })
        icut.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                if (`val` == 1) return
                changing = true
                cur.remove(thi)
                if (p1 == null) p1 = ImgCutEditPage(front, thi)
                changePanel(p1.also { cur = it!! })
                cur.add(thi)
                (cur as AbEditPage).setSelection(anim)
                `val` = 1
                changing = false
            }
        })
        mmdl.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                if (`val` == 2) return
                changing = true
                cur.remove(thi)
                if (p2 == null) p2 = MaModelEditPage(front, thi)
                changePanel(p2.also { cur = it!! })
                cur.add(thi)
                (cur as AbEditPage).setSelection(anim)
                `val` = 2
                changing = false
            }
        })
        manm.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                if (`val` == 3) return
                changing = true
                cur.remove(thi)
                if (p3 == null) p3 = MaAnimEditPage(front, thi)
                changePanel(p3.also { cur = it!! })
                cur.add(thi)
                (cur as AbEditPage).setSelection(anim)
                `val` = 3
                changing = false
            }
        })
        save.setLnr(Consumer { e: ActionEvent? -> BCUWriter.writeData() })
        undo.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                anim.restore()
                review()
                (cur as AbEditPage).setSelection(anim)
            }
        })
    }

    private fun ini() {
        add(view)
        add(icut)
        add(mmdl)
        add(manm)
        add(save)
        add(undo)
        addListeners()
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        if (`val` == 0) p0 = cur as DIYViewPage else if (`val` == 1) p1 = cur as ImgCutEditPage else if (`val` == 2) p2 = cur as MaModelEditPage else if (`val` == 3) p3 = cur as MaAnimEditPage
        ini()
    }
}
